src/server/routes/parcel.routes.ts

```diff
@@ -19,6 +21,48 @@ export default function registerRoutes(
     },
   );
 
+  async function getDeserializedParcel(payload: Buffer, logger: FastifyBaseLogger) {
+    try {
+      return await Parcel.deserialize(bufferToArrayBuffer(payload));
+    } catch (err) {
+      // Don't log the full error because 99.99% of the time the reason will suffice.
+      logger.info({ reason: (err as Error).message }, 'Refusing malformed parcel');
+      return null;
+    }
+  }
+
+  async function isMessageValid(parcel: Parcel, serviceOptions: ServiceOptions) {
+    try {
+      await serviceOptions.activeEndpoint.validateMessage(parcel);
+      return true;
+    } catch (err) {
+      serviceOptions.logger.info({ err }, 'Refusing invalid parcel');
+      return false;
+    }
+  }
+
+  async function getDecryptedParcel(parcel: Parcel, serviceOptions: ServiceOptions) {
+    try {
+      return await parcel.unwrapPayload(serviceOptions.activeEndpoint.keyStores.privateKeyStore);
+    } catch (err) {
+      serviceOptions.logger.info({ err }, 'Ignoring invalid service message');
+      return null;
+    }
+  }
+
+  async function storePda(pdaBuffer: Buffer, serviceOptions: ServiceOptions) {
+    try {
+      const privateEndpointConnParams = await PrivateEndpointConnParams.deserialize(pdaBuffer);
```

Comment:
There's no test for the malformed scenario. Make sure `savePrivateEndpointChannel` isn't called there.